ext_image_copy_capture: error "session already has a frame object"

I was testing `wl-screenrec --experiemental-ext-image-copy-capture`, from git main (specifically, b36991d5cf71aed43aed5eba0be17e7d286eaddc), against [a wlroots MR implementing ext-image-copy-capture](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4545/diffs), and found that it makes a protocol error (having two ext_image_copy_capture_frame_v1s per session at the same time) when the output being captured is resized.

The trailing part of the Wayland debug log:
```
[1399476.714][rs] -> ext_image_copy_capture_session_v1@11.create_frame(ext_image_copy_capture_frame_v1@12)
[1399476.763][rs] -> zwp_linux_dmabuf_v1@3.create_params(zwp_linux_buffer_params_v1@13)
[1399476.791][rs] -> zwp_linux_buffer_params_v1@13.add(9, 0, 0, 5120, 0, 0)
[1399476.828][rs] -> zwp_linux_buffer_params_v1@13.create_immed(wl_buffer@14, 1280, 720, 875713112, 0)
[1399476.864][rs] -> ext_image_copy_capture_frame_v1@12.attach_buffer(wl_buffer@14)
[1399476.888][rs] -> ext_image_copy_capture_frame_v1@12.capture()
[1399574.864][rs] <- wl_display@1.delete_id, (16)
[1399574.914][rs] <- wl_display@1.delete_id, (15)
[1399574.942][rs] <- wl_display@1.delete_id, (17)
[1399574.975][rs] <- ext_image_copy_capture_frame_v1@12.damage, (4, 683, 1272, 26)
[1399575.054][rs] <- ext_image_copy_capture_frame_v1@12.transform, (0)
[1399575.115][rs] <- ext_image_copy_capture_frame_v1@12.presentation_time, (0, 463847, 509870730)
[1399575.187][rs] <- ext_image_copy_capture_frame_v1@12.ready, ()
[1399575.240][rs] <- wl_buffer@14.release, ()
[1399575.338][rs] -> zwp_linux_buffer_params_v1@13.destroy()
[1399575.397][rs] -> ext_image_copy_capture_frame_v1@12.destroy()
[1399575.452][rs] -> wl_buffer@14.destroy()
[1399576.959][rs] -> ext_image_copy_capture_session_v1@11.create_frame(ext_image_copy_capture_frame_v1@15)
[1399577.022][rs] -> zwp_linux_dmabuf_v1@3.create_params(zwp_linux_buffer_params_v1@16)
[1399577.053][rs] -> zwp_linux_buffer_params_v1@16.add(9, 0, 0, 5120, 0, 0)
[1399577.116][rs] -> zwp_linux_buffer_params_v1@16.create_immed(wl_buffer@17, 1280, 720, 875713112, 0)
[1399577.154][rs] -> ext_image_copy_capture_frame_v1@15.attach_buffer(wl_buffer@17)
[1399577.179][rs] -> ext_image_copy_capture_frame_v1@15.capture()
[1399698.368][rs] <- wl_display@1.delete_id, (13)
[1399698.410][rs] <- wl_display@1.delete_id, (12)
[1399698.433][rs] <- wl_display@1.delete_id, (14)
[1399698.458][rs] <- wl_output@6.mode, (1, 1291, 720, 0)
[1399698.498][rs] <- wl_output@6.geometry, (0, 0, 0, 0, 0, Some("Unknown"), Some("Unknown"), 0)
[1399698.559][rs] <- zxdg_output_v1@7.logical_position, (0, 0)
[1399698.586][rs] <- zxdg_output_v1@7.logical_size, (1291, 720)
[1399698.613][rs] <- ext_image_copy_capture_session_v1@11.buffer_size, (1291, 720)
[1399698.640][rs] <- ext_image_copy_capture_session_v1@11.shm_format, (875713112)
[1399698.663][rs] <- ext_image_copy_capture_session_v1@11.dmabuf_device, ([128, 226, 0, 0, 0, 0, 0, 0])
[1399698.710][rs] <- ext_image_copy_capture_session_v1@11.dmabuf_format, (875713112, [<long, edited out>])
[1399698.885][rs] <- ext_image_copy_capture_session_v1@11.done, ()
[1399698.911][rs] <- wl_output@6.geometry, (0, 0, 0, 0, 0, Some("Unknown"), Some("Unknown"), 0)
[1399698.970][rs] <- wl_output@6.scale, (1)
[1399698.993][rs] <- ext_image_copy_capture_session_v1@11.buffer_size, (1291, 720)
[1399699.019][rs] <- ext_image_copy_capture_session_v1@11.shm_format, (875713112)
[1399699.044][rs] <- ext_image_copy_capture_session_v1@11.dmabuf_device, ([128, 226, 0, 0, 0, 0, 0, 0])
[1399699.091][rs] <- ext_image_copy_capture_session_v1@11.dmabuf_format, (875713112, [<long, edited out>])
[1399699.256][rs] <- ext_image_copy_capture_session_v1@11.done, ()
[1399699.277][rs] <- wl_output@6.done, ()
[1399699.493][rs] -> ext_image_copy_capture_session_v1@11.create_frame(ext_image_copy_capture_frame_v1@12)
[1399700.327][rs] -> zwp_linux_dmabuf_v1@3.create_params(zwp_linux_buffer_params_v1@13)
[1399700.372][rs] -> zwp_linux_buffer_params_v1@13.add(10, 0, 0, 5120, 0, 0)
[1399700.417][rs] -> zwp_linux_buffer_params_v1@13.create_immed(wl_buffer@14, 1280, 720, 875713112, 0)
[1399700.463][rs] -> ext_image_copy_capture_frame_v1@12.attach_buffer(wl_buffer@14)
[1399700.489][rs] -> ext_image_copy_capture_frame_v1@12.capture()
[1399700.655][rs] -> ext_image_copy_capture_session_v1@11.create_frame(ext_image_copy_capture_frame_v1@18)
[1399701.480][rs] -> zwp_linux_dmabuf_v1@3.create_params(zwp_linux_buffer_params_v1@19)
[1399701.518][rs] -> zwp_linux_buffer_params_v1@19.add(12, 0, 0, 5120, 0, 0)
[1399701.562][rs] -> zwp_linux_buffer_params_v1@19.create_immed(wl_buffer@20, 1280, 720, 875713112, 0)
[1399701.602][rs] -> ext_image_copy_capture_frame_v1@18.attach_buffer(wl_buffer@20)
[1399701.628][rs] -> ext_image_copy_capture_frame_v1@18.capture()
[1399701.853][rs] <- wl_display@1.error, (11, 1, Some("session already has a frame object"))
18:44:58 [0m[31m[ERROR] [0mProtocol error 1 on object ext_image_copy_capture_session_v1@11: session already has a frame object
thread 'main' panicked at src/main.rs:2054:45:
called `Result::unwrap()` on an `Err` value: Backend(Protocol(ProtocolError { code: 1, object_id: 11, object_interface: "ext_image_copy_capture_session_v1", message: "session already has a frame object" }))
```

--- a/src/avhw.rs
+++ b/src/avhw.rs
@@ -9,6 +9,9 @@ use ffmpeg::{
     format::Pixel,
     frame,
 };
+use log::error;
+
+use crate::DrmModifier;
 
 pub struct AvHwDevCtx {
     ptr: *mut ffmpeg::sys::AVBufferRef,
@@ -45,6 +48,7 @@ impl AvHwDevCtx {
         pixfmt: Pixel,
         width: i32,
         height: i32,
+        modifier: DrmModifier,
     ) -> Result<AvHwFrameCtx, ffmpeg::Error> {
         unsafe {
             let mut hwframe = av_hwframe_ctx_alloc(self.ptr as *mut _);
@@ -56,6 +60,10 @@ impl AvHwDevCtx {
             (*hwframe_casted).width = width;
             (*hwframe_casted).height = height;
             (*hwframe_casted).initial_pool_size = 5;
+
+            if modifier != DrmModifier::LINEAR {
+                error!("unknown how to request non-linear frames in vaapi");
+            }
 
             let sts = av_hwframe_ctx_init(hwframe);
             if sts != 0 {

--- a/src/cap_ext_image_copy.rs
+++ b/src/cap_ext_image_copy.rs
@@ -3,7 +3,7 @@ use std::path::PathBuf;
 use anyhow::Context;
 use drm::{buffer::DrmFourcc, node::DrmNode};
 use libc::dev_t;
-use log::warn;
+use log::{debug, warn};
 use log_once::warn_once;
 use wayland_client::{
     globals::GlobalList, protocol::wl_output::WlOutput, Dispatch, Proxy, QueueHandle,
@@ -20,7 +20,7 @@ use wayland_protocols::ext::{
     },
 };
 
-use crate::{CaptureSource, DmabufFormat, DmabufPotentialFormat, DrmModifier, State};
+use crate::{CaptureSource, DmabufPotentialFormat, DrmModifier, State};
 
 impl Dispatch<ExtImageCopyCaptureManagerV1, ()> for State<CapExtImageCopy> {
     fn event(
@@ -99,11 +99,7 @@ impl Dispatch<ExtImageCopyCaptureSessionV1, ()> for State<CapExtImageCopy> {
                 }
             }
             ext_image_copy_capture_session_v1::Event::Done => {
-                let mut constraints = BufferConstraints {
-                    dmabuf_formats: Vec::new(),
-                    buffer_size: None,
-                    dmabuf_device: None,
-                };
+                let mut constraints = BufferConstraints::default();
                 // All buffer constraint events will be resent on every change, so reset
                 // accumulated state
                 std::mem::swap(
@@ -114,28 +110,17 @@ impl Dispatch<ExtImageCopyCaptureSessionV1, ()> for State<CapExtImageCopy> {
                 let size = constraints
                     .buffer_size
                     .expect("Done received before BufferSize...");
-                let fmt = state.negotiate_format(
+                state.negotiate_format(
                     &constraints.dmabuf_formats,
                     size,
                     constraints.dmabuf_device.as_deref(),
+                    qhandle,
                 );
-                let Some(fmt) = fmt else {
-                    // error, it's already reported so we just have to cleanup & exit
-                    return;
-                };
-
-                let cap = state.enc.unwrap_cap();
-                cap.current_config = Some((fmt, size));
-
-                let (width, height, format, frame) = cap
-                    .queue_capture_frame(qhandle)
-                    .expect("Done without size/format!");
-                state.on_copy_src_ready(width, height, format, qhandle, &frame);
             }
             ext_image_copy_capture_session_v1::Event::Stopped => {
                 state.on_copy_fail(qhandle); // untested if this actually works
             }
-            _ => todo!(),
+            _ => {}
         }
     }
 }
@@ -151,8 +136,8 @@ impl Dispatch<ExtImageCopyCaptureFrameV1, ()> for State<CapExtImageCopy> {
     ) {
         use wayland_protocols::ext::image_copy_capture::v1::client::ext_image_copy_capture_frame_v1::Event::*;
         match event {
-            Transform { .. } => {}
-            Damage { .. } => {} // TODO: maybe this is how you implement damage
+            Transform { .. } => {} // TODO: use this
+            Damage { .. } => {}    // TODO: maybe this is how you implement damage
             PresentationTime {
                 tv_sec_hi,
                 tv_sec_lo,
@@ -162,15 +147,17 @@ impl Dispatch<ExtImageCopyCaptureFrameV1, ()> for State<CapExtImageCopy> {
                 let (hi, lo, n) = state.enc.unwrap().cap.time.take().unwrap();
                 state.on_copy_complete(qhandle, hi, lo, n);
             }
-            Failed { .. } => {
+            Failed { reason } => {
+                debug!("frame copy failed: {reason:?}");
                 state.on_copy_fail(qhandle);
             }
-            _ => {}
+            _ => todo!(),
         }
     }
 }
 
 /** Struct to collect buffer constraint information as the events arrive */
+#[derive(Default)]
 struct BufferConstraints {
     dmabuf_formats: Vec<DmabufPotentialFormat>,
     buffer_size: Option<(u32, u32)>,
@@ -181,7 +168,6 @@ pub struct CapExtImageCopy {
     output_capture_session: ExtImageCopyCaptureSessionV1,
     time: Option<(u32, u32, u32)>,
     in_progress_constraints: BufferConstraints,
-    current_config: Option<(DmabufFormat, (u32, u32))>,
 }
 
 impl CaptureSource for CapExtImageCopy {
@@ -218,28 +204,17 @@ impl CaptureSource for CapExtImageCopy {
         Ok(Self {
             output_capture_session,
             time: None,
-            in_progress_constraints: BufferConstraints {
-                dmabuf_formats: Vec::new(),
-                buffer_size: None,
-                dmabuf_device: None,
-            },
-            current_config: None,
+            in_progress_constraints: BufferConstraints::default(),
         })
     }
 
-    fn queue_capture_frame(
-        &self,
-        eq: &QueueHandle<crate::State<Self>>,
-    ) -> Option<(u32, u32, DrmFourcc, Self::Frame)> {
-        if let Some((fmt, (w, h))) = &self.current_config {
-            let frame = self.output_capture_session.create_frame(eq, ());
-            Some((*w, *h, fmt.fourcc, frame))
-        } else {
-            None
-        }
+    fn alloc_frame(&self, eq: &QueueHandle<crate::State<Self>>) -> Option<Self::Frame> {
+        debug!("ext_image_copy_capture_session_v1::create_frame");
+        let frame = self.output_capture_session.create_frame(eq, ());
+        Some(frame)
     }
 
-    fn queue_copy_frame(
+    fn queue_copy(
         &self,
         damage: bool,
         buf: &wayland_client::protocol::wl_buffer::WlBuffer,
@@ -253,6 +228,7 @@ impl CaptureSource for CapExtImageCopy {
     }
 
     fn on_done_with_frame(&self, f: Self::Frame) {
+        debug!("ext_image_copy_capture_frame_v1::destroy");
         f.destroy();
     }
 }

--- a/src/cap_ext_image_copy.rs
+++ b/src/cap_ext_image_copy.rs
@@ -67,9 +67,7 @@ impl Dispatch<ExtImageCopyCaptureSessionV1, ()> for State<CapExtImageCopy> {
     ) {
         match event {
             ext_image_copy_capture_session_v1::Event::BufferSize { width, height } => {
-                if let ExtImageCopyState::Probing(_, size, _) = &mut state.enc.unwrap_cap().state {
-                    *size = Some((width, height))
-                }
+                state.enc.unwrap_cap().in_progress_constraints.buffer_size = Some((width, height));
             }
             ext_image_copy_capture_session_v1::Event::ShmFormat { .. } => {}
             ext_image_copy_capture_session_v1::Event::DmabufDevice { device } => {
@@ -80,9 +78,7 @@ impl Dispatch<ExtImageCopyCaptureSessionV1, ()> for State<CapExtImageCopy> {
                     .unwrap()
                     .unwrap();
                 let path = node.dev_path().unwrap();
-                if let ExtImageCopyState::Probing(_, _, dev) = &mut state.enc.unwrap_cap().state {
-                    *dev = Some(path);
-                }
+                state.enc.unwrap_cap().in_progress_constraints.dmabuf_device = Some(path);
             }
             ext_image_copy_capture_session_v1::Event::DmabufFormat { format, modifiers } => {
                 assert!(modifiers.len() % 8 == 0);
@@ -92,36 +88,45 @@ impl Dispatch<ExtImageCopyCaptureSessionV1, ()> for State<CapExtImageCopy> {
                     .collect();
 
                 if let Ok(fourcc) = DrmFourcc::try_from(format) {
-                    if let ExtImageCopyState::Probing(formats, _, _) =
-                        &mut state.enc.unwrap_cap().state
-                    {
-                        formats.push(DmabufPotentialFormat { fourcc, modifiers })
-                    }
+                    state
+                        .enc
+                        .unwrap_cap()
+                        .in_progress_constraints
+                        .dmabuf_formats
+                        .push(DmabufPotentialFormat { fourcc, modifiers });
                 } else {
                     warn!("Unknown DRM Fourcc: 0x{:08x}", format)
                 }
             }
             ext_image_copy_capture_session_v1::Event::Done => {
-                // decide on format
-                let probed = if let ExtImageCopyState::Probing(formats, size, dev) =
-                    &state.enc.unwrap_cap().state
-                {
-                    Some((formats.clone(), size, dev.clone()))
-                } else {
-                    None
+                let mut constraints = BufferConstraints {
+                    dmabuf_formats: Vec::new(),
+                    buffer_size: None,
+                    dmabuf_device: None,
+                };
+                // All buffer constraint events will be resent on every change, so reset
+                // accumulated state
+                std::mem::swap(
+                    &mut state.enc.unwrap_cap().in_progress_constraints,
+                    &mut constraints,
+                );
+
+                let size = constraints
+                    .buffer_size
+                    .expect("Done received before BufferSize...");
+                let fmt = state.negotiate_format(
+                    &constraints.dmabuf_formats,
+                    size,
+                    constraints.dmabuf_device.as_deref(),
+                );
+                let Some(fmt) = fmt else {
+                    // error, it's already reported so we just have to cleanup & exit
+                    return;
                 };
 
-                if let Some((formats, size, dev)) = probed {
-                    let size = size.expect("Done received before BufferSize...");
-                    let fmt = state.negotiate_format(&formats, size, dev.as_deref());
-                    if let Some(fmt) = fmt {
-                        state.enc.unwrap_cap().state = ExtImageCopyState::Ready(fmt, size);
-                    } else {
-                        return; // error, it's already reported so we just have to cleanup & exit
-                    }
-                }
-
                 let cap = state.enc.unwrap_cap();
+                cap.current_config = Some((fmt, size));
+
                 let (width, height, format, frame) = cap
                     .queue_capture_frame(qhandle)
                     .expect("Done without size/format!");
@@ -163,19 +168,18 @@ impl Dispatch<ExtImageCopyCaptureFrameV1, ()> for State<CapExtImageCopy> {
     }
 }
 
-enum ExtImageCopyState {
-    Probing(
-        Vec<DmabufPotentialFormat>,
-        Option<(u32, u32)>,
-        Option<PathBuf>,
-    ),
-    Ready(DmabufFormat, (u32, u32)),
+/** Struct to collect buffer constraint information as the events arrive */
+struct BufferConstraints {
+    dmabuf_formats: Vec<DmabufPotentialFormat>,
+    buffer_size: Option<(u32, u32)>,
+    dmabuf_device: Option<PathBuf>,
 }
 
 pub struct CapExtImageCopy {
     output_capture_session: ExtImageCopyCaptureSessionV1,
     time: Option<(u32, u32, u32)>,
-    state: ExtImageCopyState,
+    in_progress_constraints: BufferConstraints,
+    current_config: Option<(DmabufFormat, (u32, u32))>,
 }
 
 impl CaptureSource for CapExtImageCopy {
@@ -212,7 +216,12 @@ impl CaptureSource for CapExtImageCopy {
         Ok(Self {
             output_capture_session,
             time: None,
-            state: ExtImageCopyState::Probing(Vec::new(), None, None),
+            in_progress_constraints: BufferConstraints {
+                dmabuf_formats: Vec::new(),
+                buffer_size: None,
+                dmabuf_device: None,
+            },
+            current_config: None,
         })
     }
 
@@ -220,7 +229,7 @@ impl CaptureSource for CapExtImageCopy {
         &self,
         eq: &QueueHandle<crate::State<Self>>,
     ) -> Option<(u32, u32, DrmFourcc, Self::Frame)> {
-        if let ExtImageCopyState::Ready(fmt, (w, h)) = &self.state {
+        if let Some((fmt, (w, h))) = &self.current_config {
             let frame = self.output_capture_session.create_frame(eq, ());
             Some((*w, *h, fmt.fourcc, frame))
         } else {

--- a/src/cap_ext_image_copy.rs
+++ b/src/cap_ext_image_copy.rs
@@ -162,7 +162,9 @@ impl Dispatch<ExtImageCopyCaptureFrameV1, ()> for State<CapExtImageCopy> {
                 let (hi, lo, n) = state.enc.unwrap().cap.time.take().unwrap();
                 state.on_copy_complete(qhandle, hi, lo, n);
             }
-            Failed { .. } => todo!(),
+            Failed { .. } => {
+                state.on_copy_fail(qhandle);
+            }
             _ => {}
         }
     }

--- a/src/cap_wlr_screencopy.rs
+++ b/src/cap_wlr_screencopy.rs
@@ -57,24 +57,18 @@ impl Dispatch<ZwlrScreencopyFrameV1, ()> for State<CapWlrScreencopy> {
             } => {
                 let fourcc = DrmFourcc::try_from(format).unwrap();
                 let cap = state.enc.unwrap_cap();
-                if !cap.sent_format {
-                    cap.sent_format = true;
-                    let device = cap.drm_device.clone();
-                    if state
-                        .negotiate_format(
-                            &[DmabufPotentialFormat {
-                                fourcc,
-                                modifiers: vec![DrmModifier::LINEAR],
-                            }],
-                            (dmabuf_width, dmabuf_height),
-                            device.as_deref(),
-                        )
-                        .is_none()
-                    {
-                        return; // error, which has already been reported
-                    }
-                }
-                state.on_copy_src_ready(dmabuf_width, dmabuf_height, fourcc, qhandle, capture);
+
+                let device = cap.drm_device.clone();
+                state.negotiate_format(
+                    &[DmabufPotentialFormat {
+                        fourcc,
+                        modifiers: vec![DrmModifier::LINEAR],
+                    }],
+                    (dmabuf_width, dmabuf_height),
+                    device.as_deref(),
+                    qhandle,
+                );
+                state.on_frame_allocd(qhandle, capture);
             }
             zwlr_screencopy_frame_v1::Event::Damage { .. } => {}
             zwlr_screencopy_frame_v1::Event::Buffer { .. } => {}
@@ -114,7 +108,6 @@ impl Dispatch<ZwpLinuxDmabufFeedbackV1, ()> for State<CapWlrScreencopy> {
 pub struct CapWlrScreencopy {
     screencopy_manager: ZwlrScreencopyManagerV1,
     output: WlOutput,
-    sent_format: bool,
     drm_device: Option<PathBuf>,
 }
 impl CaptureSource for CapWlrScreencopy {
@@ -134,12 +127,11 @@ impl CaptureSource for CapWlrScreencopy {
         Ok(Self {
             screencopy_manager: man,
             output,
-            sent_format: false,
             drm_device: None,
         })
     }
 
-    fn queue_copy_frame(&self, damage: bool, buf: &WlBuffer, capture: &Self::Frame) {
+    fn queue_copy(&self, damage: bool, buf: &WlBuffer, capture: &Self::Frame) {
         if damage {
             capture.copy_with_damage(buf);
         } else {
@@ -147,10 +139,7 @@ impl CaptureSource for CapWlrScreencopy {
         }
     }
 
-    fn queue_capture_frame(
-        &self,
-        eq: &QueueHandle<State<Self>>,
-    ) -> Option<(u32, u32, DrmFourcc, Self::Frame)> {
+    fn alloc_frame(&self, eq: &QueueHandle<State<Self>>) -> Option<Self::Frame> {
         // creating this triggers the linux_dmabuf event, which is where we allocate etc
 
         let _capture = self

--- a/src/main.rs
+++ b/src/main.rs
@@ -40,7 +40,7 @@ use ffmpeg::{
         AVDRMFrameDescriptor, AVPixelFormat, AV_HWFRAME_MAP_WRITE, FF_COMPLIANCE_STRICT,
     },
     filter,
-    format::{self, Pixel},
+    format::{self, Output, Pixel},
     frame::{self, video},
     media, Packet, Rational,
 };
@@ -243,11 +243,17 @@ trait CaptureSource: Sized {
         eq: &QueueHandle<State<Self>>,
         output: WlOutput,
     ) -> anyhow::Result<Self>;
-    fn queue_capture_frame(
-        &self,
-        eq: &QueueHandle<State<Self>>,
-    ) -> Option<(u32, u32, DrmFourcc, Self::Frame)>;
-    fn queue_copy_frame(&self, damage: bool, buf: &WlBuffer, cap: &Self::Frame);
+
+    // allocates a frame, either sync or async
+    // if async, return None and call `on_frame_allocd` at a later moment
+    // if sync, just return the allocated stuff
+    fn alloc_frame(&self, eq: &QueueHandle<State<Self>>) -> Option<Self::Frame>;
+
+    // queue a copy of the screen into `buf`
+    // call `on_copy_complete` or `on_copy_fail` when the copy has completed
+    fn queue_copy(&self, damage: bool, buf: &WlBuffer, cap: &Self::Frame);
+
+    // destroy the `frame` object
     fn on_done_with_frame(&self, f: Self::Frame);
 }
 
@@ -510,10 +516,12 @@ struct DmabufPotentialFormat {
     modifiers: Vec<DrmModifier>,
 }
 
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
 struct DmabufFormat {
+    width: i32,
+    height: i32,
     fourcc: DrmFourcc,
-    _modifier: DrmModifier,
+    modifier: DrmModifier,
 }
 
 #[link(name = "drm")]
@@ -523,13 +531,7 @@ extern "C" {
 }
 
 struct State<S: CaptureSource> {
-    pub(crate) surfaces_owned_by_compositor: VecDeque<(
-        frame::Video,
-        video::Video,
-        ZwpLinuxBufferParamsV1,
-        S::Frame,
-        WlBuffer,
-    )>,
+    in_flight_surface: InFlightSurface<S>,
     dma: ZwpLinuxDmabufV1,
     enc: EncConstructionStage<S>,
     starting_timestamp: Option<i64>,
@@ -539,6 +541,23 @@ struct State<S: CaptureSource> {
     sigusr1_flag: Arc<AtomicBool>,
     gm: GlobalList,
     xdg_output_manager: ZxdgOutputManagerV1,
+}
+
+enum InFlightSurface<S: CaptureSource> {
+    None,
+    AllocQueued,
+    Allocd(S::Frame),
+    CopyQueued {
+        av_surface: frame::Video,
+        av_mapping: video::Video,
+        wl_frame: S::Frame,
+        wl_buffer: WlBuffer,
+    },
+}
+impl<S: CaptureSource> InFlightSurface<S> {
+    fn take(&mut self) -> InFlightSurface<S> {
+        mem::replace(self, InFlightSurface::None)
+    }
 }
 
 struct ProbingOutputsState {
@@ -865,7 +884,7 @@ impl<S: CaptureSource + 'static> State<S> {
 
         Ok((
             State {
-                surfaces_owned_by_compositor: VecDeque::new(),
+                in_flight_surface: InFlightSurface::None,
                 dma,
                 enc: EncConstructionStage::ProbingOutputs(ProbingOutputsState {
                     partial_outputs,
@@ -883,40 +902,46 @@ impl<S: CaptureSource + 'static> State<S> {
         ))
     }
 
-    fn on_copy_src_ready(
-        &mut self,
-        dmabuf_width: u32,
-        dmabuf_height: u32,
-        format: DrmFourcc,
-        qhandle: &QueueHandle<State<S>>,
-        frame: &S::Frame,
-    ) {
+    fn on_frame_allocd(&mut self, qhandle: &QueueHandle<State<S>>, frame: &S::Frame) {
+        assert!(matches!(
+            self.in_flight_surface,
+            InFlightSurface::AllocQueued
+        ));
+        self.in_flight_surface = InFlightSurface::Allocd(frame.clone());
+
         match &mut self.enc {
             EncConstructionStage::ProbingOutputs { .. } => unreachable!(
                 "Oops, somehow created a screencopy frame without initial enc state stuff?"
             ),
             EncConstructionStage::EverythingButFormat { .. } => {
-                panic!("you need to call negotiate_format before on_copy_src_ready")
+                // queue_frame_capture will be called in negotiate_format
             }
             EncConstructionStage::OutputWentAway(_) => {
                 panic!("copy_src_ready called when the output went away??")
             }
             EncConstructionStage::Intermediate => panic!("enc left in intermediate state"),
-            EncConstructionStage::Complete(_) => {}
+            EncConstructionStage::Complete(_) => {
+                self.queue_frame_capture(qhandle);
+            }
         }
-
+    }
+    fn queue_frame_capture(&mut self, qhandle: &QueueHandle<Self>) {
         let CompleteState { enc, cap, .. } = self.enc.unwrap();
 
-        let surf = enc.frames_rgb.alloc().unwrap();
+        let InFlightSurface::Allocd(frame) = &self.in_flight_surface else {
+            panic!("queue_frame_capture called in a strange state");
+        };
 
-        let (desc, mapping) = map_drm(&surf);
+        let av_surface = enc.frames_rgb.alloc().unwrap();
+
+        let (desc, av_mapping) = map_drm(&av_surface);
 
         let modifier = desc.objects[0].format_modifier.to_be_bytes();
         let stride = desc.layers[0].planes[0].pitch as u32;
         let fd = unsafe { BorrowedFd::borrow_raw(desc.objects[0].fd) };
 
-        let dma_params = self.dma.create_params(qhandle, ());
-        dma_params.add(
+        let wl_buffer_params = self.dma.create_params(qhandle, ());
+        wl_buffer_params.add(
             fd,
             0,
             0,
@@ -925,24 +950,133 @@ impl<S: CaptureSource + 'static> State<S> {
             u32::from_be_bytes(modifier[4..].try_into().unwrap()),
         );
 
-        let buf = dma_params.create_immed(
-            dmabuf_width as i32,
-            dmabuf_height as i32,
-            format as u32,
+        let wl_buffer = wl_buffer_params.create_immed(
+            enc.selected_format.width,
+            enc.selected_format.height,
+            enc.selected_format.fourcc as u32,
             zwp_linux_buffer_params_v1::Flags::empty(),
             qhandle,
             (),
         );
 
-        cap.queue_copy_frame(self.args.damage, &buf, frame);
+        cap.queue_copy(self.args.damage, &wl_buffer, frame);
 
-        self.surfaces_owned_by_compositor.push_back((
-            surf,
-            mapping,
-            dma_params,
-            frame.clone(),
-            buf,
-        ));
+        self.in_flight_surface = InFlightSurface::CopyQueued {
+            av_surface,
+            av_mapping,
+            wl_frame: frame.clone(),
+            wl_buffer,
+        };
+    }
+
+    fn on_new_capture_format(
+        &mut self,
+        mut cs: CompleteState<S>,
+        new_format: DmabufFormat,
+    ) -> anyhow::Result<CompleteState<S>> {
+        if new_format == cs.enc.selected_format {
+            return Ok(cs);
+        }
+        info!("compositor gave new format {new_format:?}");
+
+        // destroy old frames
+        match &self.in_flight_surface {
+            InFlightSurface::Allocd(_) => {} // these frames are format independent, the previously allocated one is fine
+            InFlightSurface::CopyQueued {
+                wl_frame,
+                wl_buffer,
+                ..
+            } => {
+                cs.cap.on_done_with_frame(wl_frame.clone());
+                wl_buffer.destroy();
+                self.in_flight_surface = InFlightSurface::None;
+            }
+            InFlightSurface::None => {}
+            InFlightSurface::AllocQueued => {}
+        }
+
+        let capture_pixfmt = dmabuf_to_av(new_format.fourcc);
+
+        // make sure bounds are still valid, as size may have changed
+        cs.enc.roi_screen_coord = cs
+            .enc
+            .roi_screen_coord
+            .fit_inside_bounds(new_format.width, new_format.height);
+
+        if cs.enc.roi_screen_coord.w == 0 || cs.enc.roi_screen_coord.h == 0 {
+            bail!("new capture surface is zero-sized, bailing");
+        }
+
+        cs.enc.frames_rgb = cs.enc.hw_device_ctx
+            .create_frame_ctx(capture_pixfmt, new_format.width, new_format.height, new_format.modifier)
+            .with_context(|| format!("Failed to create vaapi frame context for capture surfaces of format {capture_pixfmt:?} {new_format:?}"))?;
+
+        // todo: proper size here
+        let enc_pixfmt_av = match cs.enc.enc_pixfmt {
+            EncodePixelFormat::Vaapi(fmt) => fmt,
+            EncodePixelFormat::Sw(fmt) => fmt,
+        };
+
+        cs.enc.selected_format = new_format;
+
+        // flush old filter & encoder
+        cs.enc
+            .video_filter
+            .get("in")
+            .unwrap()
+            .source()
+            .flush()
+            .unwrap();
+        cs.enc.process_ready();
+        if cs.enc.enc_video_has_been_fed_any_frames {
+            // ffmpeg bug--if you call send_eof before feeding any frames it will crash
+            cs.enc.enc_video.send_eof().unwrap();
+        }
+        cs.enc.process_ready();
+
+        // create a new encoder
+        // TODO: correct scaling
+        let mut frames_yuv = cs.enc.hw_device_ctx
+            .create_frame_ctx(enc_pixfmt_av, cs.enc.roi_screen_coord.w, cs.enc.roi_screen_coord.h, DrmModifier::LINEAR)
+            .with_context(|| {
+                format!("Failed to create a vaapi frame context for encode surfaces of format {enc_pixfmt_av:?} {}x{}", cs.enc.roi_screen_coord.w, cs.enc.roi_screen_coord.h)
+            })?;
+
+        let encoder = cs.enc.enc_video.codec().unwrap();
+        let framerate = cs.enc.enc_video.frame_rate();
+        let global_header = cs
+            .enc
+            .octx
+            .format()
+            .flags()
+            .contains(format::Flags::GLOBAL_HEADER);
+        let enc = make_video_params(
+            &self.args,
+            cs.enc.enc_pixfmt,
+            &encoder,
+            (cs.enc.roi_screen_coord.w, cs.enc.roi_screen_coord.h),
+            framerate,
+            global_header,
+            &mut cs.enc.hw_device_ctx,
+            &mut frames_yuv,
+        )?;
+
+        cs.enc.enc_video = enc.open_with(cs.enc.enc_video_options.clone())?;
+        cs.enc.enc_video_has_been_fed_any_frames = false;
+
+        let (filter, filter_timebase) = video_filter(
+            &mut cs.enc.frames_rgb,
+            cs.enc.enc_pixfmt,
+            (new_format.width, new_format.height),
+            cs.enc.roi_screen_coord,
+            (cs.enc.roi_screen_coord.w, cs.enc.roi_screen_coord.h),
+            cs.enc.transform,
+        );
+        cs.enc.video_filter = filter;
+        cs.enc.filter_output_timebase = filter_timebase;
+        cs.enc.format_change = true;
+
+        Ok(cs)
     }
 
     fn update_output_info_wl_output(
@@ -1015,17 +1149,13 @@ impl<S: CaptureSource + 'static> State<S> {
                         let enc = mem::replace(&mut self.enc, EncConstructionStage::Intermediate)
                             .take_enc();
                         let cap = S::new(&self.gm, qhandle, info.output.clone()).unwrap();
-                        let queue_ret = cap.queue_capture_frame(qhandle);
                         self.enc = EncConstructionStage::Complete(CompleteState {
                             enc,
                             cap,
                             output: info,
                             output_went_away: false,
                         });
-
-                        if let Some((w, h, fmt, frame)) = queue_ret {
-                            self.on_copy_src_ready(w, h, fmt, qhandle, &frame);
-                        }
+                        self.queue_alloc_frame(qhandle);
                     }
                 }
             }
@@ -1128,17 +1258,14 @@ impl<S: CaptureSource + 'static> State<S> {
                 return;
             }
         };
-
-        let queue_ret = cap.queue_capture_frame(qhandle);
+        debug!("siopc");
         self.enc = EncConstructionStage::EverythingButFormat {
             roi,
             cap,
             output: output.clone(),
         };
 
-        if let Some((w, h, fmt, frame)) = queue_ret {
-            self.on_copy_src_ready(w, h, fmt, qhandle, &frame);
-        }
+        self.queue_alloc_frame(qhandle);
     }
 
     fn on_copy_complete(
@@ -1148,17 +1275,25 @@ impl<S: CaptureSource + 'static> State<S> {
         tv_sec_lo: u32,
         tv_nsec: u32,
     ) {
+        debug!("on_copy_complete");
         let CompleteState { enc, cap, .. } = self.enc.unwrap();
 
         self.fps_counter.on_frame();
 
-        let (mut surf, drop_mapping, destroy_buffer_params, destroy_frame, destroy_buffer) =
-            self.surfaces_owned_by_compositor.pop_front().unwrap();
-
-        drop(drop_mapping);
-        destroy_buffer_params.destroy();
-        cap.on_done_with_frame(destroy_frame);
-        destroy_buffer.destroy();
+        let mut surf = if let InFlightSurface::CopyQueued {
+            av_surface,
+            av_mapping,
+            wl_frame,
+            wl_buffer,
+        } = self.in_flight_surface.take()
+        {
+            drop(av_mapping);
+            cap.on_done_with_frame(wl_frame);
+            wl_buffer.destroy();
+            av_surface
+        } else {
+            panic!("on_copy_complete called in a strange state")
+        };
 
         let secs = (i64::from(tv_sec_hi) << 32) + i64::from(tv_sec_lo);
         let pts_abs = secs * 1_000_000_000 + i64::from(tv_nsec);
@@ -1181,9 +1316,7 @@ impl<S: CaptureSource + 'static> State<S> {
 
         enc.push(surf);
 
-        if let Some((w, h, fmt, frame)) = cap.queue_capture_frame(qhandle) {
-            self.on_copy_src_ready(w, h, fmt, qhandle, &frame);
-        }
+        self.queue_alloc_frame(qhandle);
     }
 
     fn on_copy_fail(&mut self, qhandle: &QueueHandle<Self>) {
@@ -1191,16 +1324,23 @@ impl<S: CaptureSource + 'static> State<S> {
             output_went_away,
             output,
             cap,
-            ..
+            enc,
         } = self.enc.unwrap();
 
-        let (destroy_surf, drop_mapping, destroy_buffer_params, destroy_frame, destroy_buffer) =
-            self.surfaces_owned_by_compositor.pop_front().unwrap();
-        drop(drop_mapping);
-        destroy_buffer_params.destroy();
-        cap.on_done_with_frame(destroy_frame);
-        destroy_buffer.destroy();
-        drop(destroy_surf);
+        if let InFlightSurface::CopyQueued {
+            av_surface,
+            av_mapping,
+            wl_frame,
+            wl_buffer,
+        } = self.in_flight_surface.take()
+        {
+            drop(av_mapping);
+            cap.on_done_with_frame(wl_frame);
+            wl_buffer.destroy();
+            drop(av_surface);
+        } else {
+            panic!("on_copy_fail called in strange state");
+        }
 
         if *output_went_away {
             info!(
@@ -1221,11 +1361,13 @@ impl<S: CaptureSource + 'static> State<S> {
                 }
             }
             self.enc = EncConstructionStage::OutputWentAway(owa);
+        } else if enc.format_change {
+            enc.format_change = false;
+            debug!("failed transfer, but just did a format change so not surprising. trying to capture a new frame...");
+            self.queue_alloc_frame(qhandle);
         } else {
-            error!("unknown copy fail reason, trying to capture a new frame");
-            if let Some((w, h, fmt, frame)) = cap.queue_capture_frame(qhandle) {
-                self.on_copy_src_ready(w, h, fmt, qhandle, &frame);
-            }
+            error!("unknown copy fail reason, trying to capture a new frame...");
+            self.queue_alloc_frame(qhandle);
         }
     }
 
@@ -1234,8 +1376,9 @@ impl<S: CaptureSource + 'static> State<S> {
         capture_formats: &[DmabufPotentialFormat],
         (w, h): (u32, u32),
         dri_device: Option<&Path>,
-    ) -> Option<DmabufFormat> {
-        debug!("Supported capture formats are {capture_formats:?}");
+        eq: &QueueHandle<State<S>>,
+    ) {
+        debug!("Supported capture formats are {w}x{h} {capture_formats:?}");
         let dri_device = if let Some(dev) = &self.args.dri_device {
             Path::new(dev)
         } else if let Some(dev) = dri_device {
@@ -1245,43 +1388,107 @@ impl<S: CaptureSource + 'static> State<S> {
             Path::new("/dev/dri/renderD128")
         };
 
+        fn negotiate_format_impl(
+            width: i32,
+            height: i32,
+            capture_formats: &[DmabufPotentialFormat],
+        ) -> anyhow::Result<DmabufFormat> {
+            let mut selected_format = None;
+            for preferred_format in [
+                DrmFourcc::Xrgb8888,
+                DrmFourcc::Xbgr8888,
+                DrmFourcc::Xrgb2101010,
+            ] {
+                let is_fmt_supported = capture_formats.iter().any(|p| {
+                    p.fourcc == preferred_format
+                        && p.modifiers.iter().any(|m| *m == DrmModifier::LINEAR)
+                });
+
+                if is_fmt_supported {
+                    selected_format = Some(DmabufFormat {
+                        fourcc: preferred_format,
+                        modifier: DrmModifier::LINEAR,
+                        width,
+                        height,
+                    });
+                    break;
+                }
+            }
+            Ok( match selected_format {
+            Some(sf) => sf,
+            None =>
+                bail!("failed to select a viable capture format. This is probably a bug. Availabe capture formats are {:?}", capture_formats),
+        })
+        }
+
+        let selected_format = match negotiate_format_impl(w as i32, h as i32, capture_formats) {
+            Ok(f) => f,
+            Err(e) => {
+                error!("Failed to negotiate format: {e}");
+                return;
+            }
+        };
+
         match mem::replace(&mut self.enc, EncConstructionStage::Intermediate) {
             EncConstructionStage::EverythingButFormat { output, roi, cap } => {
                 let enc = match EncState::new(
                     &self.args,
-                    capture_formats,
+                    selected_format,
                     output.refresh,
                     output.transform,
-                    (w as i32, h as i32),
                     roi,
                     Arc::clone(&self.sigusr1_flag),
                     dri_device,
                 ) {
                     Ok(enc) => enc,
                     Err(e) => {
-                        error!("failed to create encoder(s): {}", e);
+                        error!("failed to create encoder(s): {e}");
                         self.quit_flag.store(1, SeqCst);
-                        return None;
+                        return;
                     }
                 };
 
-                let selected_format = enc.selected_format;
                 self.enc = EncConstructionStage::Complete(CompleteState {
                     enc,
                     cap,
                     output,
                     output_went_away: false,
                 });
-
-                Some(selected_format)
             }
-            EncConstructionStage::Complete(c) => {
-                // can happen on dispaly disconnect & reconnect
-                let selected_format = c.enc.selected_format;
+            EncConstructionStage::Complete(mut c) => {
+                // can happen on dispaly disconnect & reconnect OR output resize
+                c = match self.on_new_capture_format(c, selected_format) {
+                    Ok(enc) => enc,
+                    Err(e) => {
+                        error!("failed to renegotiate new format {selected_format:?}: {e}");
+                        self.quit_flag.store(1, SeqCst);
+                        return;
+                    }
+                };
                 self.enc = EncConstructionStage::Complete(c);
-                Some(selected_format)
             }
             _ => panic!("called negotiate_format in a strange state"),
+        }
+
+        // make the next sensible step in capture
+        match &self.in_flight_surface {
+            InFlightSurface::None => {
+                self.queue_alloc_frame(eq);
+            }
+            InFlightSurface::AllocQueued => {} // nothing to do
+            InFlightSurface::Allocd(_) => {
+                self.queue_frame_capture(eq);
+            }
+            InFlightSurface::CopyQueued { .. } => {}
+        }
+    }
+
+    fn queue_alloc_frame(&mut self, eq: &QueueHandle<State<S>>) {
+        assert!(matches!(self.in_flight_surface, InFlightSurface::None));
+        let f = self.enc.unwrap_cap().alloc_frame(eq);
+        self.in_flight_surface = InFlightSurface::AllocQueued;
+        if let Some(f) = f {
+            self.on_frame_allocd(eq, &f);
         }
     }
 }
@@ -1289,6 +1496,7 @@ impl<S: CaptureSource + 'static> State<S> {
 struct EncState {
     video_filter: filter::Graph,
     enc_video: encoder::Video,
+    enc_video_has_been_fed_any_frames: bool,
     octx: format::context::Output,
     frames_rgb: AvHwFrameCtx,
     filter_output_timebase: Rational,
@@ -1297,6 +1505,12 @@ struct EncState {
     sigusr1_flag: Arc<AtomicBool>,
     audio: Option<AudioHandle>,
     selected_format: DmabufFormat,
+    hw_device_ctx: AvHwDevCtx,
+    enc_pixfmt: EncodePixelFormat,
+    roi_screen_coord: Rect,
+    transform: Transform,
+    enc_video_options: dictionary::Owned<'static>,
+    format_change: bool,
 }
 
 #[derive(Copy, Clone, Debug)]
@@ -1361,7 +1575,7 @@ fn make_video_params(
     Ok(enc)
 }
 
-fn parse_dict(dict: &str) -> Result<dictionary::Owned, ffmpeg::Error> {
+fn parse_dict<'a>(dict: &str) -> Result<dictionary::Owned<'a>, ffmpeg::Error> {
     let cstr = CString::new(dict).unwrap();
 
     let mut ptr = null_mut();
@@ -1369,8 +1583,8 @@ fn parse_dict(dict: &str) -> Result<dictionary::Owned, ffmpeg::Error> {
         let res = av_dict_parse_string(
             &mut ptr,
             cstr.as_ptr(),
-            b"=:\0".as_ptr().cast(),
-            b",\0".as_ptr().cast(),
+            c"=:".as_ptr().cast(),
+            c",".as_ptr().cast(),
             0,
         );
         if res != 0 {
@@ -1381,14 +1595,82 @@ fn parse_dict(dict: &str) -> Result<dictionary::Owned, ffmpeg::Error> {
     }
 }
 
+fn get_encoder(args: &Args, format: &Output) -> anyhow::Result<ffmpeg::Codec> {
+    Ok(if let Some(encoder_name) = &args.ffmpeg_encoder {
+        ffmpeg_next::encoder::find_by_name(encoder_name).ok_or_else(|| {
+            format_err!(
+                "Encoder {encoder_name} specified with --ffmpeg-encoder could not be instntiated"
+            )
+        })?
+    } else {
+        let codec_id = match args.codec {
+            Codec::Auto => format.codec(&args.filename, media::Type::Video),
+            Codec::Avc => codec::Id::H264,
+            Codec::Hevc => codec::Id::HEVC,
+            Codec::VP8 => codec::Id::VP8,
+            Codec::VP9 => codec::Id::VP9,
+            Codec::AV1 => codec::Id::AV1,
+        };
+
+        let maybe_hw_codec = if args.hw {
+            if let Some(hw_codec_name) = vaapi_codec_id(codec_id) {
+                if let Some(codec) = ffmpeg_next::encoder::find_by_name(hw_codec_name) {
+                    Some(codec)
+                } else {
+                    warn!("there is a known vaapi codec ({hw_codec_name}) for codec {codec_id:?}, but it's not available. Using a generic encoder...");
+                    None
+                }
+            } else {
+                warn!("hw flag is specified, but there's no known vaapi codec for {codec_id:?}. Using a generic encoder...");
+                None
+            }
+        } else {
+            None
+        };
+
+        match maybe_hw_codec {
+            Some(codec) => codec,
+            None => match ffmpeg_next::encoder::find(codec_id) {
+                Some(codec) => codec,
+                None => {
+                    bail!("Failed to get any encoder for codec {codec_id:?}");
+                }
+            },
+        }
+    })
+}
+
+fn get_enc_pixfmt(args: &Args, encoder: &ffmpeg::Codec) -> anyhow::Result<EncodePixelFormat> {
+    let supported_formats = supported_formats(encoder);
+    Ok(if supported_formats.is_empty() {
+        match args.encode_pixfmt {
+            Some(fmt) => EncodePixelFormat::Sw(fmt),
+            None => {
+                warn!(
+                    "codec \"{}\" does not advertize supported pixel formats, just using NV12. Pass --encode-pixfmt to suppress this warning",
+                    encoder.name()
+                );
+                EncodePixelFormat::Sw(Pixel::NV12)
+            }
+        }
+    } else if supported_formats.contains(&Pixel::VAAPI) {
+        EncodePixelFormat::Vaapi(args.encode_pixfmt.unwrap_or(Pixel::NV12))
+    } else {
+        match args.encode_pixfmt {
+            None => EncodePixelFormat::Sw(supported_formats[0]),
+            Some(fmt) if supported_formats.contains(&fmt) => EncodePixelFormat::Sw(fmt),
+            Some(fmt) => bail!("Encoder does not support pixel format {fmt:?}"),
+        }
+    })
+}
+
 impl EncState {
     // assumed that capture_{w,h}
     fn new(
         args: &Args,
-        capture_formats: &[DmabufPotentialFormat],
+        capture_format: DmabufFormat,
         refresh: Rational,
         transform: Transform,
-        (capture_w, capture_h): (i32, i32), // pixels
         roi_screen_coord: Rect, // roi in screen coordinates (0, 0 is screen upper left, which is not necessarily captured frame upper left)
         sigusr1_flag: Arc<AtomicBool>,
         dri_device: &Path,
@@ -1405,98 +1687,12 @@ impl EncState {
             ffmpeg_next::format::output_with(&args.filename, muxer_options).unwrap()
         };
 
-        let encoder = if let Some(encoder_name) = &args.ffmpeg_encoder {
-            ffmpeg_next::encoder::find_by_name(encoder_name).ok_or_else(|| {
-                format_err!(
-                    "Encoder {encoder_name} specified with --ffmpeg-encoder could not be instntiated"
-                )
-            })?
-        } else {
-            let codec_id = match args.codec {
-                Codec::Auto => octx.format().codec(&args.filename, media::Type::Video),
-                Codec::Avc => codec::Id::H264,
-                Codec::Hevc => codec::Id::HEVC,
-                Codec::VP8 => codec::Id::VP8,
-                Codec::VP9 => codec::Id::VP9,
-                Codec::AV1 => codec::Id::AV1,
-            };
-
-            let maybe_hw_codec = if args.hw {
-                if let Some(hw_codec_name) = vaapi_codec_id(codec_id) {
-                    if let Some(codec) = ffmpeg_next::encoder::find_by_name(hw_codec_name) {
-                        Some(codec)
-                    } else {
-                        warn!("there is a known vaapi codec ({hw_codec_name}) for codec {codec_id:?}, but it's not available. Using a generic encoder...");
-                        None
-                    }
-                } else {
-                    warn!("hw flag is specified, but there's no known vaapi codec for {codec_id:?}. Using a generic encoder...");
-                    None
-                }
-            } else {
-                None
-            };
-
-            match maybe_hw_codec {
-                Some(codec) => codec,
-                None => match ffmpeg_next::encoder::find(codec_id) {
-                    Some(codec) => codec,
-                    None => {
-                        bail!("Failed to get any encoder for codec {codec_id:?}");
-                    }
-                },
-            }
-        };
+        let encoder = get_encoder(args, &octx.format())?;
 
         // format selection: naive version, should actually see what the ffmpeg filter supports...
-        let mut selected_format = None;
-        for preferred_format in [
-            DrmFourcc::Xrgb8888,
-            DrmFourcc::Xbgr8888,
-            DrmFourcc::Xrgb2101010,
-        ] {
-            let is_fmt_supported = capture_formats.iter().any(|p| {
-                p.fourcc == preferred_format
-                    && p.modifiers.iter().any(|m| *m == DrmModifier::LINEAR)
-            });
+        info!("capture pixel format is {}", capture_format.fourcc);
 
-            if is_fmt_supported {
-                selected_format = Some(DmabufFormat {
-                    fourcc: preferred_format,
-                    _modifier: DrmModifier::LINEAR,
-                });
-                break;
-            }
-        }
-        let selected_format = match selected_format {
-            Some(sf) => sf,
-            None =>
-                bail!("failed to select a viable capture format. This is probably a bug. Availabe capture formats are {:?}", capture_formats),
-        };
-        let capture_pixfmt = dmabuf_to_av(selected_format.fourcc);
-        info!("capture pixel format is {}", selected_format.fourcc);
-
-        let supported_formats = supported_formats(&encoder);
-        let enc_pixfmt = if supported_formats.is_empty() {
-            match args.encode_pixfmt {
-                Some(fmt) => EncodePixelFormat::Sw(fmt),
-                None => {
-                    warn!(
-                        "codec \"{}\" does not advertize supported pixel formats, just using NV12. Pass --encode-pixfmt to suppress this warning",
-                        encoder.name()
-                    );
-                    EncodePixelFormat::Sw(Pixel::NV12)
-                }
-            }
-        } else if supported_formats.contains(&Pixel::VAAPI) {
-            EncodePixelFormat::Vaapi(args.encode_pixfmt.unwrap_or(Pixel::NV12))
-        } else {
-            match args.encode_pixfmt {
-                None => EncodePixelFormat::Sw(supported_formats[0]),
-                Some(fmt) if supported_formats.contains(&fmt) => EncodePixelFormat::Sw(fmt),
-                Some(fmt) => bail!("Encoder does not support pixel format {fmt:?}"),
-            }
-        };
+        let enc_pixfmt = get_enc_pixfmt(args, &encoder)?;
         info!("encode pixel format is {enc_pixfmt:?}");
 
         let codec_id = encoder.id();
@@ -1536,8 +1732,8 @@ impl EncState {
         };
 
         let mut frames_rgb = hw_device_ctx
-            .create_frame_ctx(capture_pixfmt, capture_w, capture_h)
-            .with_context(|| format!("Failed to create vaapi frame context for capture surfaces of format {capture_pixfmt:?} {capture_w}x{capture_h}"))?;
+            .create_frame_ctx(dmabuf_to_av(capture_format.fourcc), capture_format.width, capture_format.height, capture_format.modifier)
+            .with_context(|| format!("Failed to create vaapi frame context for capture surfaces of format {capture_format:?}"))?;
 
         let (enc_w_screen_coord, enc_h_screen_coord) = match args.encode_resolution {
             Some((x, y)) => (x as i32, y as i32),
@@ -1547,7 +1743,7 @@ impl EncState {
         let (video_filter, filter_timebase) = video_filter(
             &mut frames_rgb,
             enc_pixfmt,
-            (capture_w, capture_h),
+            (capture_format.width, capture_format.height),
             roi_screen_coord,
             (enc_w_screen_coord, enc_h_screen_coord),
             transform,
@@ -1558,9 +1754,9 @@ impl EncState {
             EncodePixelFormat::Sw(fmt) => fmt,
         };
         let mut frames_yuv = hw_device_ctx
-            .create_frame_ctx(enc_pixfmt_av, enc_w_screen_coord, enc_h_screen_coord)
+            .create_frame_ctx(enc_pixfmt_av, enc_w_screen_coord, enc_h_screen_coord, DrmModifier::LINEAR)
             .with_context(|| {
-                format!("Failed to create a vaapi frame context for encode surfaces of format {enc_pixfmt_av:?} {capture_w}x{capture_h}")
+                format!("Failed to create a vaapi frame context for encode surfaces of format {enc_pixfmt_av:?} {enc_w_screen_coord}x{enc_h_screen_coord}")
             })?;
 
         info!("{}", video_filter.dump());
@@ -1581,7 +1777,7 @@ impl EncState {
             None => dict!(),
         };
 
-        let enc_video = if args.hw {
+        let (enc_video, enc_video_options) = if args.hw {
             let low_power_opts = {
                 let mut d = passed_enc_options.clone();
                 d.set("low_power", "1");
@@ -1597,32 +1793,35 @@ impl EncState {
             };
 
             match args.low_power {
-                LowPowerMode::Auto => match enc.open_with(low_power_opts) {
-                    Ok(enc) => enc,
+                LowPowerMode::Auto => match enc.open_with(low_power_opts.clone()) {
+                    Ok(enc) => (enc, low_power_opts),
                     Err(e) => {
                         eprintln!("failed to open encoder in low_power mode ({}), trying non low_power mode. if you have an intel iGPU, set enable_guc=2 in the i915 module to use the fixed function encoder. pass --low-power=off to suppress this warning", e);
-                        make_video_params(
-                            args,
-                            enc_pixfmt,
-                            &encoder,
-                            (enc_w_screen_coord, enc_h_screen_coord),
-                            refresh,
-                            global_header,
-                            &mut hw_device_ctx,
-                            &mut frames_yuv,
-                        )?
-                        .open_with(regular_opts)?
+                        (
+                            make_video_params(
+                                args,
+                                enc_pixfmt,
+                                &encoder,
+                                (enc_w_screen_coord, enc_h_screen_coord),
+                                refresh,
+                                global_header,
+                                &mut hw_device_ctx,
+                                &mut frames_yuv,
+                            )?
+                            .open_with(regular_opts.clone())?,
+                            regular_opts,
+                        )
                     }
                 },
-                LowPowerMode::On => enc.open_with(low_power_opts)?,
-                LowPowerMode::Off => enc.open_with(regular_opts)?,
+                LowPowerMode::On => (enc.open_with(low_power_opts.clone())?, low_power_opts),
+                LowPowerMode::Off => (enc.open_with(regular_opts.clone())?, regular_opts),
             }
         } else {
             let mut enc_options = passed_enc_options.clone();
             if enc_options.get("preset").is_none() {
                 enc_options.set("preset", "ultrafast");
             }
-            enc.open_with(enc_options).unwrap()
+            (enc.open_with(enc_options.clone()).unwrap(), enc_options)
         };
 
         let mut ost_video = octx.add_stream(encoder).unwrap();
@@ -1651,14 +1850,21 @@ impl EncState {
         Ok(EncState {
             video_filter,
             enc_video,
+            enc_video_has_been_fed_any_frames: false,
             filter_output_timebase: filter_timebase,
             octx,
             vid_stream_idx,
+            hw_device_ctx,
+            enc_pixfmt,
+            roi_screen_coord,
+            transform,
+            enc_video_options,
             frames_rgb,
             history_state,
             sigusr1_flag,
             audio,
-            selected_format,
+            selected_format: capture_format,
+            format_change: false,
         })
     }
 
@@ -1709,6 +1915,7 @@ impl EncState {
         {
             // encoder has same time base as the filter, so don't do any time scaling
             self.enc_video.send_frame(&yuv_frame).unwrap();
+            self.enc_video_has_been_fed_any_frames = true;
         }
 
         let mut encoded = Packet::empty();
@@ -1858,7 +2065,6 @@ fn video_filter(
     transform: Transform,
 ) -> (filter::Graph, Rational) {
     let mut g = ffmpeg::filter::graph::Graph::new();
-
 
     // src
     unsafe {

--- a/src/main.rs
+++ b/src/main.rs
@@ -1190,8 +1190,17 @@ impl<S: CaptureSource + 'static> State<S> {
         let CompleteState {
             output_went_away,
             output,
+            cap,
             ..
         } = self.enc.unwrap();
+
+        let (destroy_surf, drop_mapping, destroy_buffer_params, destroy_frame, destroy_buffer) =
+            self.surfaces_owned_by_compositor.pop_front().unwrap();
+        drop(drop_mapping);
+        destroy_buffer_params.destroy();
+        cap.on_done_with_frame(destroy_frame);
+        destroy_buffer.destroy();
+        drop(destroy_surf);
 
         if *output_went_away {
             info!(
@@ -1213,8 +1222,10 @@ impl<S: CaptureSource + 'static> State<S> {
             }
             self.enc = EncConstructionStage::OutputWentAway(owa);
         } else {
-            error!("unknown copy fail reason, quitting");
-            self.quit_flag.store(1, SeqCst)
+            error!("unknown copy fail reason, trying to capture a new frame");
+            if let Some((w, h, fmt, frame)) = cap.queue_capture_frame(qhandle) {
+                self.on_copy_src_ready(w, h, fmt, qhandle, &frame);
+            }
         }
     }
 

--- a/src/transform.rs
+++ b/src/transform.rs
@@ -83,6 +83,20 @@ impl Rect {
             h: (y1 - y2).abs(),
         }
     }
+
+    pub fn bottom_right(&self) -> (i32, i32) {
+        (self.x + self.w, self.y + self.h)
+    }
+
+    #[must_use]
+    pub fn fit_inside_bounds(&self, w: i32, h: i32) -> Rect {
+        let mut r = *self;
+        r.x = 0.max(w.min(r.bottom_right().0) - r.w);
+        r.y = 0.max(h.min(r.bottom_right().1) - r.h);
+        r.w = r.w.min(w);
+        r.h = r.h.min(h);
+        r
+    }
 }
 
 #[cfg(test)]
@@ -92,6 +106,16 @@ mod test {
     use crate::transform::transform_is_transposed;
 
     use super::Rect;
+
+    #[test]
+    fn fit_inside_bounds() {
+        let r = Rect::new((10, 10), (20, 20));
+
+        assert_eq!(r.fit_inside_bounds(30, 30), r);
+        assert_eq!(r.fit_inside_bounds(20, 30), Rect::new((0, 10), (20, 20)));
+        assert_eq!(r.fit_inside_bounds(30, 20), Rect::new((10, 0), (20, 20)));
+        assert_eq!(r.fit_inside_bounds(10, 5), Rect::new((0, 0), (10, 5)));
+    }
 
     #[test]
     fn screen_to_frame_normal() {

